sound/soc/sof/compress.c

```diff
@@ -0,0 +1,42 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// Copyright 2021 NXP
+//
+// Author: Daniel Baluta <daniel.baluta@nxp.com>
+
+#include <sound/soc.h>
+#include <sound/sof.h>
+#include <sound/compress_driver.h>
+#include "sof-audio.h"
+#include "sof-priv.h"
+
+void snd_sof_compr_fragment_elapsed_work(struct work_struct *work)
+{
+	struct snd_sof_pcm_stream *sps =
+		container_of(work, struct snd_sof_pcm_stream,
+			     period_elapsed_work);
+
+	snd_compr_fragment_elapsed(sps->cstream);
+}
+
+/*
+ * sof compr fragment elapse, this could be called in irq thread context
+ */
+void snd_sof_compr_fragment_elapsed(struct snd_compr_stream *cstream)
+{
+	struct snd_soc_pcm_runtime *rtd = cstream->private_data;
+	struct snd_soc_component *component =
+		snd_soc_rtdcom_lookup(rtd, SOF_AUDIO_PCM_DRV_NAME);
+	struct snd_sof_pcm *spcm;
+
+	spcm = snd_sof_find_spcm_dai(component, rtd);
+	if (!spcm) {
+		dev_err(component->dev,
+			"error: fragment elapsed for unknown stream!\n");
+		return;
+	}
+
+	/* see snd_sof_pcm_period_elapsed */
```

Comment:
snd_sof_compr_fragment_elapsed?

Reply:
@plbossart actually this comment was meant to send the reader to snd_sof_pcm_period_elapsed in order to read this comment. Should, I just copy this comment here also?

```
       /*
»        * snd_pcm_period_elapsed() can be called in interrupt context                                                                                                                                     
»        * before IRQ_HANDLED is returned. Inside snd_pcm_period_elapsed(),
»        * when the PCM is done draining or xrun happened, a STOP IPC will
»        * then be sent and this IPC will hit IPC timeout.
»        * To avoid sending IPC before the previous IPC is handled, we
»        * schedule delayed work here to call the snd_pcm_period_elapsed().
»        */
```

Reply:
sorry, I misunderstood the comment.

Maybe reword it as 'use the same workqueue-based solution as for PCM, c.f. snd_sof_pcm_elapsed'

Reply:
@plbossart I always asked people to make code more clear to understand. I'm getting old :)